tools/nut-scanner/scan_eaton_serial.c

```diff
@@ -425,6 +436,19 @@ nutscan_device_t * nutscan_scan_eaton_serial(const char* ports_range)
 		nutscan_scan_eaton_serial_device(current_port_name);
 #endif
 		current_port_nb++;
+            } else {
+#ifdef HAVE_PTHREAD
+                if (thread_array != NULL) {
+                    for ( i=0; i < thread_count ; i++) {
+                        pthread_join(thread_array[i],NULL);
+                        sem_post(semaphore);
+                    }
+                    thread_count = 0;
+                    free(thread_array);
+                    thread_array = NULL;
+                }
+#endif
+            }
```

Comment:
I do not quite understand the logic. From a quick look, we are decrementing the semaphore each time a thread is created and as soon as the semaphore reaches 0, we skip the current port (?!), join all threads and only then we create new threads? If the intent was to rate-limit the thread creation, a simple blocking sem_wait() when starting a thread and sem_post() when joining it would do the job.

Reply:
I stand corrected, we do not skip the port (current_port_nb is only incremented inside the iteration). So from a functional perspective, it's correct, but still rather complicated if the intent was to rate-limit thread creation.